File: vol2/ch21/LeetCode847.cpp
class Solution {
    const int INF=0x3f3f3f3f;
    vector<vector<int>> A;
public:
    int shortestPathLength(vector<vector<int>>& graph) {
        int n=graph.size();
        A=vector<vector<int>>(n,vector<int>(n,INF));
        for(int i=0;i<n;i++) {      //建立邻接矩阵A
            for(int j:graph[i]) A[i][j]=1;
        }
        Floyd();
        int m=1<<n;           //总的状态数
        vector<vector<int>> dp=vector<vector<int>>(m,vector<int>(n,INF));
        for (int i=0;i<n;i++) dp[1<<i][i]=0;
        for (int state=0;state<m;state++) { //枚举所有的state
            for (int i=0;i<n;i++) {     			//枚举state中已访问的顶点i
                if(!inset(state,i)) continue;  //顶点i不在state中时跳过
                for(int j=0;j<n;j++) {      //枚举state中尚未访问过的顶点j
                    if (inset(state,j)) continue;   //顶点j在state中时跳过
                    int state1=addj(state,j);
                    dp[state1][j]=min(dp[state1][j],dp[state][i]+A[i][j]);
                }
            }
        }
        int ans=INF;
        for (int i=0;i<n;i++)
            ans=min(ans,dp[m-1][i]);
        return ans;
    }
    bool inset(int state,int j) {        //判断顶点j是否在state中
        return (state & (1<<j))!=0;
    }
    int addj(int state,int j) {             //在state添加顶点j
        return state | (1<<j);
    }

    void Floyd() {
        int n=A.size();
        for(int k=0;k<n;k++) {
            for (int i=0;i<n;i++) {
                for (int j=0;j<n;j++)
                    A[i][j]=min(A[i][j],A[i][k]+A[k][j]);
            }
        }
    }
};

File: vol2/ch21/LeetCode847.py
class Solution:
    def shortestPathLength(self, graph: List[List[int]]) -> int:
        INF=0x3f3f3f3f
        n=len(graph)
        self.A=[[INF for _ in range(n)] for _ in range(n)]
        for i in range(0,n):    #建立邻接矩阵A
            for j in graph[i]:self.A[i][j]=1
        self.Floyd()
        m=1<<n;           #总的状态数
        dp=[[INF for _ in range(n)] for _ in range(m)]
        for i in range(0,n):dp[1<<i][i]=0
        for state in range(0,m):     #枚举所有的state
            for i in range(0,n):    #枚举state中已访问的i
                if not self.inset(state,i):continue;  #i不在state中时跳过
                for j in range(0,n):    #枚举state中尚未访问过的j
                    if self.inset(state,j):continue  #j在state中时跳过
                    state1=self.addj(state,j)
                    dp[state1][j]=min(dp[state1][j],dp[state][i]+self.A[i][j])
        return min(dp[m-1])
    def inset(self,state,j):       #判断顶点j是否在state中
        return (state & (1<<j))!=0
    def addj(self,state,j):       #在state添加顶点j
        return state | (1<<j)
    def Floyd(self):
        n=len(self.A)
        for k in range(0,n):
            for i in range(0,n):
                for j in range(0,n):
                    self.A[i][j]=min(self.A[i][j],self.A[i][k]+self.A[k][j])
